### PR TITLE
Add missing log levels

### PR DESCRIPTION
**Pull request by SUPERCILEX**
I'd like to be able to default to `TraceLevel` when a tracing feature is enabled in my crate.

### CODE REVIEW

**Comment by epage** [comment]
Our current assumptions:
- Debug and trace by definition are below the threshold for normal operations and we should encourage best practices of defaults being for normal operations
- Abnormal cases where debug and trace become normal operations that people can workaround it the above by providing their own `LogLevel` impl

Is there a reason we should re-think these assumptions?  Otherwise, I lean to closing this with the specified workaround.

**Comment by SUPERCILEX** [comment]
This isn't changing the default, it's giving developers the freedom to choose what works for them. I gave you my use case, I want to be able to do this:

```rust
#[cfg(not(feature = "trace"))]
type DefaultLevel = clap_verbosity_flag::WarnLevel;
#[cfg(feature = "trace")]
type DefaultLevel = clap_verbosity_flag::TraceLevel;
```

Your reply is essentially telling me, "I don't like your choice so eff off." Very frustrating.

I understand that I can just make my own `TraceLevel` struct, but that doesn't make any sense. The library integrates with the log crate, so it should mirror the log crate's levels. And again, the default is error and that's fine. I'm simply asking for the *option* to choose something else without it being a pain.

Let me put it another way: by your reasoning, if I'm vegetarian and I think being vegetarian is a "best practice", then I'll tell anyone who wants to buy meat to eff off and raise their own cattle. Nonsense.

**Comment by SUPERCILEX** [comment]
What's frustrating here is that I get the sense there's nothing I can say that will change your mind. Your reply presents an option to argue against something that clearly isn't refutable. Obviously "Debug and trace by definition are below the threshold for normal operations" so the reply is an insincere pretense of collaboration and that really irks me.

The actual question here is _should devs be able to make their own choices?_ and the answer "no" is insulting. Are other devs really too dumb to decide their defaults? Making everyone suffer to get the behavior they want because you don't like it is unacceptable.

---

**One option would be to make them `doc(hidden)`.**

---

Apologies for getting pissed off yet again, but I don't think I've ***EVER*** had an interaction with you that isn't immediately met with rebuttal/rejection/confrontation. It's maddening.

**Comment by epage** [comment]
> What's frustrating here is that I get the sense there's nothing I can say that will change your mind. Your reply presents an option to argue against something that clearly isn't refutable.

I gave the base assumptions and opened a dialog for re-evaluating them.   Instead of a good-faith dialog, I am yet again being met with hostility at accepting your contributions at face value.  This type of dialog should be happening in an issue but I let us past here hoping to smooth things out.

This is your warning: if you can't interact with others without hostility, you will be banned from this and related Orgs.  Seeking better understanding is not a hostile act that deserves hostility being returned but is an expected part of being a maintainer.

> The actual question here is should devs be able to make their own choices? and the answer "no" is insulting. Are other devs really too dumb to decide their defaults? Making everyone suffer to get the behavior they want because you don't like it is unacceptable.

The answer is not "no" but "its already supported by implementing `LogLevel`".  This follows the principle of API design of make the right / safe choice easy while making the exception cases possible.  What I'm trying to understand is why does the impl need to be in here, rather than in your crate.  Bringing up use cases like having a feature flag to control the default is very relevant information that was left out of the PR.  It would be of interest to include more of that use case, how representative that is, and, if its rare, any are extenuating circumstances for why these other impls should be shipped by default.

**Comment by SUPERCILEX** [comment]
> What I'm trying to understand is why does the impl need to be in here, rather than in your crate.

Simply because I don't want to have to copy that impl in all of my crates.

> accepting your contributions at face value

No, I'm asking for two things:
- Acknowledging that my request is reasonable. I'm always happy to change the implementation or write code in a way that makes the maintainer happy, but a baseline for collaboration is acknowledging that somebody else needs something you don't.
- Compromise: for example, making the structs `doc(hidden)` is a good path forward. It meets my need for not having to duplicate code everywhere and it meets your need for making it more difficult to go against best practices.

---

> Bringing up use cases like having a feature flag to control the default is very relevant information that was left out of the PR.

Lol no? I'll quote myself: "I'd like to be able to **default** to TraceLevel when ***a tracing feature*** is **enabled** in my crate."

> It would be of interest to include more of that use case, how representative that is, and, if its rare, any are extenuating circumstances for why these other impls should be shipped by default.

This is once again disingenuous because a data driven approach doesn't make sense when the request is as dumb as "plz mirror the log crate's levels so I don't have to duplicate code."

---

A note: you're the only person I've had issues with like this in what's now going to be almost a decade of being on GitHub.

**Comment by SUPERCILEX** [comment]
> This follows the principle of API design of make the right / safe choice easy while making the exception cases possible.

This is the part I fundamentally don't understand. Typically one would describe the difference between a high level API and low level API like this: you write low level APIs that are difficult to use correctly but afford complete control (in this case implementing `LogLevel`) and then you write the common use case high level APIs that are easy to use on top of the low level ones (in this case `*Level` structs).

Ok, but why do people split the API like this instead of only providing the most flexible API?
- Performance: high level APIs typically make fewer assumptions (to keep their use simple) and thus must put in extra work to bring the program into a known state. Thus, low level APIs might be necessary to help devs meet performance targets.
- Maintenance burden: trying to provide all the possible high level use cases might be too difficult, so you only provide the common ones and tell people to use the low level API for highly specific needs.
- \*Convenience/hiding complexity: you might be able to provide high level APIs with no performance loss that meet a common need. The goal is just to hide all the low level boiler plate for the common case.
- Changing implementation details: if the low-level API exists but can't be used, then typically the high level API serves as an abstraction barrier to enable changing the low level APIs indiscriminately.
- ???

---

In this case, there's nothing performance related and there isn't a maintenance burden (maybe if this was the stdlib you'd need a bit more rational, but the fact that there are already similar APIs for the other log levels should make this a pretty easy argument). There also isn't an issue around getting stuck with an implementation you don't like.

\*I'm asking for the convenience and minimal complexity of using a high level API.

I believe you're saying one of the reasons people perform a low/high level API split is because the high level API is supposed to be endorsed by the maintainer while the low level API is the "you're on your own because it's a bad idea" path. I very strongly disagree with this take.

If I'm understanding your reasoning correctly, then again I feel like there's nothing I can do to change your mind, so what's the point? My view is that if something is simple to implement, performant, has no maintenance burden, and is something people want, then it almost always makes sense to implement it. The burden of proof should be on "why not?" and "because I don't want to let people use a default I think is bad for the ecosystem" to be valid reasoning.

Disagreement is fine, but usually people will try to compromise to make progress. Your original reply can be destructed as saying the following:
- These are my assumptions.
- I made a decision using those assumptions.
- Unless you can invalidate my assumptions, I'm sticking with my decision.

How am I supposed to respond? Of course if this was our first interaction, I would elaborate and try to explain my side of things assuming you'd want to help me make progress, but I consistently get these ultimatums from you that leave no room for compromise. What am I supposed to do?

**Comment by joshka** [comment]
Hey, @epage I saw this got a little heated above, and while I don't want to stir that up again, I'd like to throw a use case into the mix. If you'd prefer me to open another issue rather than continuing this one, or comment in #95 instead, I'm happy to do that.

TL;DR: This got a little long while writing, but I'm not sure what to cut out:
- I'd find the extra log levels useful, especially during prototyping and for situations where for a small amount of time setting the default level is an easier approach to debug some  problem at a more verbose default log level.
- Not doing this has caused confusing fragmentation that seems worse than the doing it

---

During prototyping a cli app, it would be nice and let me fall into the pit of success to have a default level of debug available to me. This would make it easy for me to have http requests show up in my logs and various things like that. Once I'm done prototyping this can certainly be switched back to the default.

If I'm writing a few different CLIs for various things, this additional boilerplate code to set a default that is not available seems heavy. It ends up getting duplicated across a few places, and it's effectively throw away code that I want to delete when it's no longer in use. But I also want the ability to turn that back on if needs be.

So given these concerns:
- wanting a simple one line setup while prototyping to avoid writing much code for simple things
- wanting the ability to change this simply in a symmetric way that aligns with changing from ERROR to INFO as the default
- not wanting to have to dig into the implementation of this to work out which parts to implement

I come to the same conclusion that including the other levels is the right thing based on those assumptions.

Right now there are 3 similarly named crates. The changes aren't documented, but I think they're:
- clap-verbosity-flag this lib
- clap-verbosity a fork which adds serde support
- clap-verbosity-flag2 a fork which adds the extra levels

As a consumer looking to simplify a verbosity flag in a cli app, this is a confusing place to be in. I want both of those extra features, but I also want to generally use software which I know will get regularly maintained rather than being a single issue fork.

I think much of what I wrote seems to align with the arguments that @SUPERCILEX was making here.

>  This follows the principle of API design of make the right / safe choice easy while making the exception cases possible

I think this way to define how an API should choose features has some problems, as it overindexes on what is "correct". I don't think there is a correct universally accepted log level that's most useful for all use cases. For some use cases, that level is DEBUG, for some it's ERROR. I don't particularly know any specific use cases where logging at a TRACE level by default makes sense, but I can see that it's reasonable to want that.

Here, there seems like there's a reasonable default (ERROR) that aligns with a default that's used in many logging frameworks, and that works well enough, but not providing the other levels seems just odd and the sort of thing that only has downside.

Perhaps there's a specific use case that you're thinking of that's unstated here where defaulting the verbosity flag level to DEBUG / TRACE would cause known issues? I wonder if you could expand on those a bit more, because those seem like the more theoretical problems to me (but it's likely I don't have the same perspective as yours, so I'm seeking to understand yours a bit more here too).
